### DISCUSSION TITLE
How to run a managed object reference function?

### DISCUSSION THREAD

**Discussion by SomniVertix** [vmware/govmomi#3224] [Community Help] [answered]
Looking to use the following method from the vsphere API:

https://vdc-download.vmware.com/vmwb-repository/dcr-public/b50dcbbf-051d-4204-a3e7-e1b618c1e384/538cf2ec-b34f-4bae-a332-3820ef9e7773/vim.host.NetworkSystem.html#queryNetworkHint

It says that its called from the "HostNetworkSystem" managed object, but whenever i try that it errors saying that it doesnt have that function.

Example code:
```go
	Run(func(ctx context.Context, c *vim25.Client) error {
		// Create a view of HostSystem objects
		m := view.NewManager(c)

		v, err := m.CreateContainerView(ctx, c.ServiceContent.RootFolder, []string{"HostNetworkSystem"}, true)
		if err != nil {
			return err
		}

		defer v.Destroy(ctx)

		// Retrieve summary property for all hosts
		// Reference: http://pubs.vmware.com/vsphere-60/topic/com.vmware.wssdk.apiref.doc/vim.HostNetworkSystem.html
		var hns []mo.HostNetworkSystem
		err = v.Retrieve(ctx, []string{"HostNetworkSystem"}, []string{}, &hns)
		if err != nil {
			return err
		}

		for _, hn := range hns {
			hn.QueryNetworkHint()
		}

		return nil
	})
```

**Comment by dougm** [comment, marked as answer]
You can't use HostNetworkSystem with [ContainerView](https://vdc-download.vmware.com/vmwb-repository/dcr-public/184bb3ba-6fa8-4574-a767-d0c96e2a38f4/ba9422ef-405c-47dd-8553-e11b619185b2/SDK/vsphere-ws/docs/ReferenceGuide/vim.view.ViewManager.html#createContainerView)
> type - An optional list of managed entity types

Where [managed entity](https://vdc-download.vmware.com/vmwb-repository/dcr-public/184bb3ba-6fa8-4574-a767-d0c96e2a38f4/ba9422ef-405c-47dd-8553-e11b619185b2/SDK/vsphere-ws/docs/ReferenceGuide/vim.ManagedEntity.html) is VirtualMachine, HostSystem, etc.

[HostNetworkSystem](https://vdc-download.vmware.com/vmwb-repository/dcr-public/184bb3ba-6fa8-4574-a767-d0c96e2a38f4/ba9422ef-405c-47dd-8553-e11b619185b2/SDK/vsphere-ws/docs/ReferenceGuide/vim.host.NetworkSystem.html) is a property of [HostConfigManager](https://vdc-download.vmware.com/vmwb-repository/dcr-public/184bb3ba-6fa8-4574-a767-d0c96e2a38f4/ba9422ef-405c-47dd-8553-e11b619185b2/SDK/vsphere-ws/docs/ReferenceGuide/vim.host.ConfigManager.html), which is a property of [HostSystem](https://vdc-download.vmware.com/vmwb-repository/dcr-public/184bb3ba-6fa8-4574-a767-d0c96e2a38f4/ba9422ef-405c-47dd-8553-e11b619185b2/SDK/vsphere-ws/docs/ReferenceGuide/vim.HostSystem.html#field_detail). And HostSystem is managed entity.

The generated `mo.HostNetworkSystem` does have methods attached to it (likewise for other `mo` types), but you can use the `object` package to simplify fetching of the properties you need and invoking the method:

```go
	simulator.Run(func(ctx context.Context, c *vim25.Client) error {
		m := view.NewManager(c)
		kind := []string{"HostSystem"}

		v, err := m.CreateContainerView(ctx, c.ServiceContent.RootFolder, []string{"HostSystem"}, true)
		if err != nil {
			log.Fatal(err)
		}

		refs, err := v.Find(ctx, kind, nil)
		if err != nil {
			return err
		}

		for _, ref := range refs {
			hn, err := object.NewHostSystem(c, ref).ConfigManager().NetworkSystem(ctx)
			if err != nil {
				return err
			}
			info, err := hn.QueryNetworkHint(ctx, nil)
			if err != nil {
				return err
			}
			_ = info // ...
		}

		return v.Destroy(ctx)
	})
```
